File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<int>dp(n,0);
        dp[0] =0;
        int mini = prices[0];
        for(int i=1;i<n;i++){
            // if(prices[i]>mini){
                dp[i] = max(dp[i-1],prices[i]-mini);
            // }else{
            //     dp[i] = dp[i-1];
            // }
            mini = min(prices[i],mini);
        }
        return dp[n-1];
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dp =[0] * n
        mini = prices[0]
        if(n<=1): 
            return 0
        dp[0] = max(0,prices[1]-prices[0])
        for i in range(1,n):
            dp[i] = max(dp[i-1] , prices[i]-mini)
            mini = min(mini,prices[i])
        return dp[n-1]
        